### PR TITLE
feat: create_segments

### PR DESCRIPTION
**Pull request by kobayu858**
## Description

- Creation of segments
- Creation of parameters for later use

## Related links

- original paper [OPEN_LINK](https://drive.google.com/file/d/1YH-MAWsw1VAvxqco4Beg3njQ9ykG-sb-/view?usp=drive_link) p4 Fig2 (b)
- paper summary [EmbIV_INTERNAL_LINK](https://docs.google.com/presentation/d/1rabhhA9ziIcBjvT8XPLue78XEWcKMvH4/edit?usp=drive_link&ouid=114908281052440997053&rtpof=true&sd=true) p23
- design specifications [EmbIV INTERNAL LINK](https://docs.google.com/presentation/d/14YO4wHY1QqGOjoY9_g4djvtTaBXV0RCj/edit?usp=drive_link&ouid=114908281052440997053&rtpof=true&sd=true) p11, p12
## Pre-review checklist for the PR author

- [x] I've confirmed the [self-checklist](https://docs.google.com/spreadsheets/d/1p5KMcm752iCySHJz0xL1qf6fcLDKprbo/edit#gid=863618628).
- [x] The PR is ready for review.

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR has been properly tested.
- [x] The PR has been reviewed.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions.
- [x] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

### CODE REVIEW

File: 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    Heavy,
+    Light,
+}
+
+pub struct Segment {
+    pub nodes: Vec<NodeData>,
+    pub begin_range: i32,
+    pub end_range: i32,
+    pub deadline: f32,                         //TODO: use
+    pub classification: SegmentClassification, //TODO: use
```
Comment (atsushi421) [inline 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs:15]:
```suggestion
    pub classification: Option<SegmentClassification>, //TODO: use
```

File: 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        } else {
+            earliest_finish_times[i - 1]
+        };
+
+        let segment = Segment {
+            nodes: Vec::new(),
+            begin_range,
+            end_range: earliest_finish_times[i],
+            deadline: 0.0,
+            classification: SegmentClassification::Heavy,
```
Comment (atsushi421) [inline 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs:45]:
```suggestion
            classification: None,
```

File: 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    dag.calculate_earliest_finish_times();
+
+    let mut earliest_finish_times = Vec::new();
+
+    for node in dag.node_weights_mut() {
+        earliest_finish_times.push(node.params["earliest_finish_time"]);
+    }
+
+    earliest_finish_times.sort();
+    earliest_finish_times.dedup();
```
Comment (atsushi421) [inline 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs:29]:
```suggestion
    earliest_finish_times.dedup();
    earliest_finish_times.sort();
```

File: 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        let node = dag.node_weight(node_i).unwrap();
+
+        for segment in &mut segments {
+            if node.params["earliest_start_time"] <= segment.begin_range
+                && segment.end_range <= node.params["earliest_finish_time"]
+            {
+                segment.nodes.push(node.clone());
+            }
+        }
+    }
```
Comment (atsushi421) [inline 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs:61]:
```suggestion
    for node in dag.node_weights() {
        for segment in &mut segments {
            if node.params["earliest_start_time"] <= segment.begin_range
                && segment.end_range <= node.params["earliest_finish_time"]
            {
                segment.nodes.push(node.clone());
            }
        }
    }
```

**Review by atsushi421 — CHANGES_REQUESTED** [review]

File: 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs (hunk 4 above)
Comment (kobayu858) [reply 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs]: Thank you very much. Correction.

File: 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs (hunk 3 above)
Comment (kobayu858) [reply 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs]: Certainly this one is faster to sort.
Fixed.

File: 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs (hunk 2 above)
Comment (kobayu858) [reply 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs]: Correction.

File: 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs (hunk 1 above)
Comment (kobayu858) [reply 2014_TPDS_basic_decomposition_based_algorithm/src/handle_segment.rs]: Correction.

**Review by atsushi421 — APPROVED** [review]
LGTM

**Pull request merged**